File: backend/java/localpos-user-service/src/main/java/com/pds/localpos/userservice/controller/UserController.java
package com.pds.localpos.userservice.controller;

import com.pds.localpos.userservice.dto.request.UserRequest;
import com.pds.localpos.userservice.dto.response.UserResponse;
import com.pds.localpos.userservice.service.UserService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@Slf4j
@RestController
@RequestMapping("/api/v1/users")
@RequiredArgsConstructor
public class UserController {

    private final UserService userService;

    @PreAuthorize("hasAnyRole('OWNER','ADMIN')")
    @GetMapping
    public ResponseEntity<List<UserResponse>> getAllUsers() {
        log.info("GET /users - Fetching all users");
        return ResponseEntity.ok(userService.getAllUsers());
    }

    @PreAuthorize("hasAnyRole('OWNER','ADMIN','MANAGER')")
    @GetMapping("/{id}")
    public ResponseEntity<UserResponse> getUser(@PathVariable String id) {
        log.info("GET /users/{} - Fetching user by ID", id);
        return ResponseEntity.ok(userService.getUserById(id));
    }

    @PreAuthorize("hasAnyRole('OWNER','ADMIN','MANAGER')")
    @PostMapping
    public ResponseEntity<UserResponse> createUser(@Valid @RequestBody UserRequest dto) {
        log.info("POST /users - Creating user: {}", dto.username());
        return ResponseEntity.ok(userService.createUser(dto));
    }

    @PreAuthorize("hasAnyRole('OWNER','ADMIN','MANAGER')")
    @PutMapping("/{id}")
    public ResponseEntity<UserResponse> updateUser(@PathVariable String id, @Valid @RequestBody UserRequest dto) {
        log.info("PUT /users/{} - Updating user", id);
        return ResponseEntity.ok(userService.updateUser(id, dto));
    }

    @PreAuthorize("hasAnyRole('OWNER','ADMIN')")
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteUser(@PathVariable String id) {
        log.warn("DELETE /users/{} - Deleting user", id);
        userService.deleteUser(id);
        return ResponseEntity.noContent().build();
    }
}

File: backend/kotlin/localpos-user-service-kt/src/main/kotlin/com/pds/localpos/userservice/controller/UserController.kt
package com.pds.localpos.userservice.controller

import com.pds.localpos.userservice.dto.LoginRequestDTO
import com.pds.localpos.userservice.dto.UserRequestDTO
import com.pds.localpos.userservice.dto.UserResponseDTO
import com.pds.localpos.userservice.mapper.toResponseDTO
import com.pds.localpos.userservice.service.UserService
import jakarta.validation.Valid
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1/users")
class UserController(
    private val userService: UserService,
    private val passwordEncoder: PasswordEncoder
) {

    private val logger = LoggerFactory.getLogger(UserController::class.java)

    @PreAuthorize("hasAnyRole('OWNER','ADMIN')")
    @GetMapping
    fun getAllUsers(): ResponseEntity<List<UserResponseDTO>> {
        logger.info("Fetching all users")
        return ResponseEntity.ok(userService.getAllUsers())
    }

    @PreAuthorize("hasAnyRole('OWNER','ADMIN','MANAGER')")
    @GetMapping("/{id}")
    fun getUser(@PathVariable id: String): ResponseEntity<UserResponseDTO> {
        logger.info("Fetching user with ID: {}", id)
        return ResponseEntity.ok(userService.getUserById(id))
    }

    @PreAuthorize("hasAnyRole('OWNER','ADMIN','MANAGER')")
    @PostMapping
    fun createUser(@Valid @RequestBody dto: UserRequestDTO): ResponseEntity<UserResponseDTO> {
        logger.info("Creating new user with username: {}", dto.username)
        return ResponseEntity.ok(userService.createUser(dto))
    }

    @PreAuthorize("hasAnyRole('OWNER','ADMIN','MANAGER')")
    @PutMapping("/{id}")
    fun updateUser(
        @PathVariable id: String,
        @Valid @RequestBody dto: UserRequestDTO
    ): ResponseEntity<UserResponseDTO> {
        logger.info("Updating user with ID: {}", id)
        return ResponseEntity.ok(userService.updateUser(id, dto))
    }

    @PreAuthorize("hasAnyRole('OWNER','ADMIN')")
    @DeleteMapping("/{id}")
    fun deleteUser(@PathVariable id: String): ResponseEntity<Void> {
        logger.warn("Deleting user with ID: {}", id)
        userService.deleteUser(id)
        return ResponseEntity.noContent().build()
    }

    @PostMapping("/auth/validate")
    fun validateUser(@RequestBody request: LoginRequestDTO): ResponseEntity<UserResponseDTO> {
        logger.info("Authenticating user: {}", request.username)

        val user = userService.findByUsername(request.username)
        if (user == null || !passwordEncoder.matches(request.password, user.password)) {
            logger.warn("Authentication failed for user: {}", request.username)
            throw BadCredentialsException("Invalid credentials")
        }

        logger.info("Authentication successful for user: {}", request.username)
        return ResponseEntity.ok(user.toResponseDTO())
    }
}
